### PR TITLE
Change compaction retention time granularity to hour

### PR DESCRIPTION
**Pull request by zzmao**
Compaction needs to run with less retention time in some use cases. Change compaction retention time granularity to hour from day.

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/StorageManager.java [hunk 1]
```diff
@@ -165,8 +165,8 @@ private void verifyConfigs(StoreConfig storeConfig, DiskManagerConfig diskManage
     /* NOTE: We must ensure that the store never performs hard deletes on the part of the log that is not yet flushed.
        We do this by making sure that the retention period for deleted messages (which determines the end point for hard
        deletes) is always greater than the log flush period. */
-    if (storeConfig.storeDeletedMessageRetentionDays
-        < TimeUnit.SECONDS.toDays(storeConfig.storeDataFlushIntervalSeconds) + 1) {
+    if (storeConfig.storeDeletedMessageRetentionHours
+        < TimeUnit.SECONDS.toHours(storeConfig.storeDataFlushIntervalSeconds) + 1) {
       throw new StoreException("Message retention days must be greater than the store flush interval period",
```
Comment (SophieGuo410) [inline ambry-store/src/main/java/com/github/ambry/store/StorageManager.java:170]: minor: Might want to change to "Message retention hours" for the exception message.

**Review by SophieGuo410 — APPROVED** [review]
LGTM.

File: ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java [hunk 2]
```diff
@@ -648,7 +648,7 @@ private boolean preUndeleteValidation(CloudBlobMetadata metadata, StoreKey key,
       throw new StoreException("Id " + key + " is already undeleted in cloud", StoreErrorCodes.ID_Undeleted);
     } else if (!metadata.isDeleted()) {
       throw new StoreException("Id " + key + " is not deleted yet in cloud ", StoreErrorCodes.ID_Not_Deleted);
-    } else if (metadata.getDeletionTime() + TimeUnit.DAYS.toMillis(storeConfig.storeDeletedMessageRetentionDays)
+    } else if (metadata.getDeletionTime() + TimeUnit.DAYS.toMillis(storeConfig.storeDeletedMessageRetentionHours)
```
Comment (jsjtzyy) [inline ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java:651]: `TimeUnit.HOURS.toMillis`

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 3]
```diff
@@ -247,7 +247,7 @@ public void start() throws StoreException {
             diskIOScheduler, metrics, time, sessionId, storeDescriptor.getIncarnationId());
         compactor.initialize(index);
         metrics.initializeIndexGauges(storeId, index, capacityInBytes);
-        long logSegmentForecastOffsetMs = TimeUnit.DAYS.toMillis(config.storeDeletedMessageRetentionDays);
+        long logSegmentForecastOffsetMs = TimeUnit.HOURS.toMillis(config.storeDeletedMessageRetentionHours);
         long bucketSpanInMs = TimeUnit.MINUTES.toMillis(config.storeStatsBucketSpanInMinutes);
```
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:251]: @justinlin-linkedin if `logSegmentForecastOffsetMs` becomes a few hours and `bucketSpanInMs`, `bucketCount` stay unchanged, will it be a problem? `logSegmentForecastEndTimeMsForDeleted`  > scan timestamp

Comment (zzmao) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:251]: @justinlin-linkedin  any comment on Ying's concern?

Comment (justinlin-linkedin) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:251]: @jsjtzyy @zzmao  I don't think so, buckets are designed to serve the incoming requests, it shouldn't be affected by the retention duration. One we get a smooth transition is to not enable bucket reporting in delivery cluster.

Comment (zzmao) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:251]: Thanks and noted. I will disable bucket reporting in EI testing.

Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:251]: Sure, we can temporarily disable bucket and revisit this before turning it on.

**Review by jsjtzyy — APPROVED** [review]

File: ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java [hunk 4]
```diff
@@ -75,11 +75,11 @@
...
 
   /**
-   * How long (in days) a key must be in deleted state before it is hard deleted.
+   * How long (in hours) a key must be in deleted state before it is hard deleted. Minimum value: 1 hour.
    */
-  @Config("store.deleted.message.retention.days")
-  @Default("7")
-  public final int storeDeletedMessageRetentionDays;
+  @Config("store.deleted.message.retention.hours")
+  @Default("24")
```
Comment (jsjtzyy) [inline ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java:81]: One minor suggestion:  maybe we can use 7 * 24 as default value which lines up with we are using for main cluster, and change this for delivery cluster only.

Comment (zzmao) [reply ambry-api/src/main/java/com/github/ambry/config/StoreConfig.java:81]: sure.

**Pull request merged**